src/shared/tpm2-util.h

```diff
@@ -66,7 +66,25 @@ static inline void tpm2_context_freep(struct tpm2_context **c) {
                 *c = tpm2_context_free(*c);
 }
 
-ESYS_TR tpm2_flush_context_verbose(struct tpm2_context *c, ESYS_TR handle);
+struct tpm2_handle {
+        struct tpm2_context *context;
+        ESYS_TR handle;
+};
+
+#define _tpm2_handle_init(c, h) (struct tpm2_handle) { .context = (c), .handle = (h), }
+#define tpm2_handle_init(c) _tpm2_handle_init(c, ESYS_TR_NONE)
+struct tpm2_handle tpm2_handle_release(struct tpm2_handle h);
+static inline void tpm2_handle_releasep(struct tpm2_handle *h) {
+        if (h)
+                *h = tpm2_handle_release(*h);
+}
+
+#define HANDLE_NONE tpm2_handle_init(NULL)
+#define HANDLE_RH_OWNER _tpm2_handle_init(NULL, ESYS_TR_RH_OWNER)
+#define HANDLE_PASSWORD _tpm2_handle_init(NULL, ESYS_TR_PASSWORD)
+
+/* Like TAKE_PTR() but for struct tpm2_handle, resetting them to HANDLE_NONE */
+#define TAKE_HANDLE(handle) TAKE_GENERIC(handle, struct tpm2_handle, HANDLE_NONE)
 
 void tpm2_pcr_mask_to_selection(uint32_t mask, uint16_t bank, TPML_PCR_SELECTION *ret);
```

Comment:
so, i wonder, this means if you have a function where you allocate a tpm2_context and then also a tpm2_handle, it is essential that you declare the former first, and the latter second. Because otherwise we'l try to reference a context that doesn't exist anymore...

hence I was wondering: maybe make the context ref-counted? that would make things super easy: the handles would pin the context, and all would be good.

adding ref counting would be really simple, we have DEFINE_TRIVIAL_REF_FUNC() macro that generates the ref/unref functions for you if you just supply a destructor function that shall be called once the ref count hits zero.

Reply:
excellent idea, i'll add that, thanks!